### ISSUE TITLE
Re-enable `variants::disk_v2::tests::size_limits::writer_waits_when_buffer_is_full`

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#11825]
Ignored in https://github.com/vectordotdev/vector/pull/11815 since it was causing the test suites to hang in CI sometimes.

It might be worth surveying the rest of the tests is this suite to see what their failure behavior is. Ideally they'd timeout and fail rather than waiting for conditions that may never be hit.

**Comment by tobz** [comment]
Yeah, maybe there's a way for the test runner to be configured on a per-test basis or something?

It's not very easy to instrument every single call site where we might be effectively waiting for an operation to complete.

**Comment by jszwedko** [comment]
Spotted another test that failed over here:

https://github.com/vectordotdev/vector/runs/5629119518?check_suite_focus=true

`reader_throws_error_when_record_has_decoding_error`

As part of this work it'd be nice to make sure all of these tests time out.

**Comment by tobz** [comment]
I'll see if I can use that `ntest::timeout` macro to wrap entire tests.  The code seems straightforward, but I just don't know if it will also correctly compose with the `tokio::test` macro itself.
